tree/treeplayer/inc/ROOT/TTreeProcessorMT.hxx

```diff
@@ -87,6 +87,8 @@ namespace ROOT {
                fChain->Add(fileNames[i].c_str(), nEntries[i]);
             }
             fChain->ResetBit(TObject::kMustCleanup);
+            // Each task reads one cluster, so we don't need to pre-fetch anything else.
+            fChain->SetCacheSize(0);
```

Comment:
This is a no go.  We must NOT disable caching.  This will destroy performance over high latency links.

Reply:
I don't understand: this is _only_ for multi-thread runs where each TTree is used to read one cluster. Is that cluster not fetched all together at once over network anyway?

Reply:
Would it be better to call `SetCacheEntryRange` when spawning a task on a cluster instead of `SetCacheSize` here?

Reply:
We need to find balance between "no cache" and "read too much per task".  The performance penalty to read N MB when asking a single cluster is also a no go. 
The problem I believe is that the current caching mechanism is not adequate for a MT execution context. Is there a way to cache only what one will read, for example setting event ranges?

Reply:
> Would it be better to call SetCacheEntryRange when spawning a task on a cluster instead of SetCacheSize here?

My understanding is that `SetCacheEntryRange` has no effect when reading through `TTreeReader`, correct @JavierCVilla ?

Reply:
> `SetCacheEntryRange` has no effect when reading through `TTreeReader`

That's correct, indeed it performs worse than using the cache without `SetCacheEntryRange`.

Bytes read asking for one entry from one single cluster with `TTreeReader`:

- No options: 102.116 bytes
- Only with `SetCacheEntryRange(start, end+1)`: 25.234.622 bytes
- Only with `SetCacheSize(0)`: 67.819 bytes (size of the cluster)

Reply:
The number with `SetCacheEntryRange(start, end+1)` (25.234.622) and TTreeReader is exactly what we were getting with TChain too before Philippe pushed his last fixes. This means we need a similar fix for TTreeReader.

Reply:
> I don't understand: this is only for multi-thread runs where each TTree is used to read one cluster. Is that cluster not fetched all together at once over network anyway?

The 'fetching a cluster in one read' is **only** done by the TTreeCache.  When not using the cache (SetCacheSize(0)), each individual basket is read by itself (so 32K or less network packets and in some cases many of them ....)

Reply:
> Is there a way to cache only what one will read, for example setting event ranges?

Yes, that whats SetCacheEntryRange is supposed to do.  The other dimensions is how many branches it needs to read.  When in 'training mode', the TTreeCache has (essentially) no choice but to read all the branch for the '1st' cluster.

Reply:
>    SetCacheEntryRange has no effect when reading through TTreeReader
> That's correct, indeed it performs worse than using the cache without SetCacheEntryRange.
> Bytes read asking for one entry from one single cluster with TTreeReader:
>    No options: 102.116 bytes
>    Only with SetCacheEntryRange(start, end+1): 25.234.622 bytes
>    Only with SetCacheSize(0): 67.819 bytes (size of the cluster)

That is in part because the training range and the cache entry range are non overlapping.  (not enough details but it sounds like all the branches are read int he 'bad' case.

Reply:
> My understanding is that SetCacheEntryRange has no effect when reading through TTreeReader, correct 

That sounds like an over-generalization :)

Reply:
Note: I am confused of what actually circumstances are __still__ broken (especially since the change for the 'TChain' case should help also TTreeReader unless we are not comparing apple to apple (i.e. different training setups).

Reply:
@JavierCVilla perhaps you can add to ROOT-9773 a (minimal) reproducer with the TChain case and the TTreeReader case, which shows how the former case is fixed but we still suffer from the issue in the latter case?

Reply:
@pcanal ok I understand way less than I thought about ROOT I/O...I wish these things were documented somewhere? :sweat_smile:

@JavierCVilla has 5 lines that reproduce the extra reading with `SetCacheEntryRange`+`TTreeReader`.

Reply:
I posted the reproducer in the [jira issue ROOT-9773](https://sft.its.cern.ch/jira/browse/ROOT-9773?focusedCommentId=87791&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-87791).